### ISSUE TITLE
Mock wallet backend

### ISSUE THREAD

**Issue by j-mueller** [input-output-hk/plutus#1337]
The mock wallet backend is a mockchain with HTTP interface that simulates wallets interacting with the blockchain (powered by the emulator). It lets us test Plutus contracts in the app platform without needing a test net / real wallet.

The mock wallet backend needs to maintain two pieces of state:

1. One `Wallet.Emulator.Types.WalletState` per simulated wallet
2. One `Wallet.Emulator.TypesEmulatorState` for the blockchain

It should have HTTP endpoints for
* balancing a transaction (`Language.Plutus.Contract.Wallet.handleTx`)
* adding an address to the wallet's list of interesting addresses, and querying for new transactions at the interesting addresses
* asking for the current slot number

All of the functionality is already implemented somewhere, so this task is mostly about figuring out the right interface.

**Comment by raduom** [comment]
Suggested API:

```
listen :: ContractId
       -> Slot
       -- ^ The slot where we start listening for the Hook.
       -> Hook
       -- ^ The hooks are: watch address and wait for slot
       -> ()

runBlocks :: ContractId
          -> Slot
          -- ^ The slot where we start streaming events.
          -> Either Slot (Event, Slot)
          -- ^ There may be no trigger up to the current slot
          --   (left branch of the Either) or we have an event
          --   that needs processing, so we suspend streaming
          --   at _snd . fromRight_

balanceTx :: UnbalancedTx
          -> Error String Tx
          -- ^ We may be unable to balance the transaction.

currentSlot :: Slot
```

**Comment by j-mueller** [comment]
Seems reasonable, just a few clarifications

* Does `ContractId` refer to an instance of the contract? If so then it will have to be generated by the app platform
* Will the mock wallet delete events after they have been consumed by 'runBlocks'?

**Comment by michaelpj** [comment]
I'm slightly unclear about the architecture: I think what happens is:
- The SCB is going to use this interface.
- The SCB will probably have two threads for interaction with the wallet backend:
    - One which calls `runBlocks` so it can respond to notifications from the wallet backend
    - One which responds to user input and calls other things

I'm also not sure what a `ContractId` is. The wallet doesn't know about contracts, those are something the SCB knows about if anything. I think we could get away with just having ids for listeners, in which case they should probably be created by the wallet backend and returned when you register a listener.

What about signing and submitting transactions? That's the wallet's job.

`runBlocks` seems like an odd name me, since that doesn't really explain what we're getting back, which is a notification. I'd call it something like `await`.

When will `runBlocks` ever return `Left`?

There are two approaches to listeners and events which I think Jann and I discussed before in some old comment thread:
- Registering and deregistering listeners are their own actions, awaiting just gets an event and doesn't do anything to the listeners.
    - Possibly this makes more sense in a multi-threaded model, since we may want to add more listeners after we call `runBlocks` but before it returns, e.g. in response to user input.
- Listeners are removed as soon as they receive an event
    - If you want to listen to many events you have to resubscribe the listener each time.
    - The advantage is that you just need a single function, which takes the listener and returns when it triggers, also removing the listener.

**Comment by raduom** [comment]
> Seems reasonable, just a few clarifications
> 
> * Does `ContractId` refer to an instance of the contract? If so then it will have to be generated by the app platform

It is just an identifier I use in `runBlocks`, it can really be anything, but it should identify a contract, hence the name :)

> * Will the mock wallet delete events after they have been consumed by 'runBlocks'?

I would like to say no. This would be a hidden state and that makes life more complicated. At the same time I don't think we can offer any guarantees since wallet may not have all the blockchain downloaded.

**Comment by raduom** [comment]
> I'm slightly unclear about the architecture: I think what happens is:
> 
> * The SCB is going to use this interface.
> * The SCB will probably have two threads for interaction with the wallet backend:
>   
>   * One which calls `runBlocks` so it can respond to notifications from the wallet backend
>   * One which responds to user input and calls other things
> 
> I'm also not sure what a `ContractId` is. The wallet doesn't know about contracts, those are something the SCB knows about if anything. I think we could get away with just having ids for listeners, in which case they should probably be created by the wallet backend and returned when you register a listener.
> 
> What about signing and submitting transactions? That's the wallet's job.
> 
> `runBlocks` seems like an odd name me, since that doesn't really explain what we're getting back, which is a notification. I'd call it something like `await`.
> 
> When will `runBlocks` ever return `Left`?
> 
> There are two approaches to listeners and events which I think Jann and I discussed before in some old comment thread:
> 
> * Registering and deregistering listeners are their own actions, awaiting just gets an event and doesn't do anything to the listeners.
>   
>   * Possibly this makes more sense in a multi-threaded model, since we may want to add more listeners after we call `runBlocks` but before it returns, e.g. in response to user input.
> * Listeners are removed as soon as they receive an event
>   
>   * If you want to listen to many events you have to resubscribe the listener each time.
>   * The advantage is that you just need a single function, which takes the listener and returns when it triggers, also removing the listener.
> I'm slightly unclear about the architecture: I think what happens is:
> 
> * The SCB is going to use this interface.
> * The SCB will probably have two threads for interaction with the wallet backend:
>   
>   * One which calls `runBlocks` so it can respond to notifications from the wallet backend
>   * One which responds to user input and calls other things
> 
> I'm also not sure what a `ContractId` is. The wallet doesn't know about contracts, those are something the SCB knows about if anything. I think we could get away with just having ids for listeners, in which case they should probably be created by the wallet backend and returned when you register a listener.

That is what ContractId is. It identifies the sets of hooks we listen for when `runBlocks` is called.

> What about signing and submitting transactions? That's the wallet's job.

This is one of the points where I think we may want to discuss a bit. Right now we give a lot of control to the users in the way they build their transactions, balance, etc. I was thinking of merging balance + fee management + signing + sending in one method. If we want this then there should only be one `handleTx` endpoint, if not we will probably need to add more.

> `runBlocks` seems like an odd name me, since that doesn't really explain what we're getting back, which is a notification. I'd call it something like `await`.

I do not have any preference here. However, note that this method does not await, it's more like a filter / search over the blocks, starting from start block looking for events matching specified hooks. What about `scanBlocks`?

> When will `runBlocks` ever return `Left`?

If from the specified slot up to the current slot there have been no events matching the conditions specified by the hooks.

> There are two approaches to listeners and events which I think Jann and I discussed before in some old comment thread:
> 
> * Registering and deregistering listeners are their own actions, awaiting just gets an event and doesn't do anything to the listeners.
>   
>   * Possibly this makes more sense in a multi-threaded model, since we may want to add more listeners after we call `runBlocks` but before it returns, e.g. in response to user input.
> * Listeners are removed as soon as they receive an event
>   
>   * If you want to listen to many events you have to resubscribe the listener each time.
>   * The advantage is that you just need a single function, which takes the listener and returns when it triggers, also removing the listener.

We are going for the second approach, but we do have an inherently asynchronous model here. So we need to functions. One that specified the listener (adds the hooks - `listen`) and one that scans the chain up to the current slot for matches (`runBlocks`).

**Comment by michaelpj** [comment]
> That is what ContractId is. It identifies the sets of hooks we listen for when runBlocks is called.

I think it would be cleaner to have one id per hook, rather than one per "set of hooks".

> This is one of the points where I think we may want to discuss a bit. Right now, we give a lot of control to the users in the way they build their transactions, balance, etc. 

"Users" now means the wallet, since we're talking about the wallet backend, right? I think it's okay for the wallet to have lots of control.

Having said that, I immediately realised that if we're talking about the wallet backend then it *doesn't* sign transactions, nor does it balance transactions (that's the *wallet*'s job). I'm now confused again: which API is this?

> I do not have any preference here. However, note that this method does not await, it's more like a filter/search over the blocks, starting from start block looking for events matching specified hooks. What about scanBlocks?

Ah I see, we have a fundamental disagreement about whether we're doing polling or callbacks.

I'd assumed we were doing callbacks (which is what the current design does), so you register a listener and a callback with the wallet backend, and it calls you when the listener triggers (maybe never, or maybe there's a timeout, or who knows).

The alternative is that the wallet has to poll the backend to find out if anything has happened, which I think is your solution.

My (limited) understanding is that:
- If the server has lots of clients, polling can be better since callbacks requires the server to do more.
    - Our server (the wallet backend) has only one client.
-If the client cares more about its own schedule then polling is better, if it cares more about responding quickly to events then callbacks are better.
    - We care more about responding quickly, I think.
- Polling systems are easier to shut down and start up again.
    - We do care about this, but I think we can work something out.

We should also check with the wallet folks what kind of interface they're expecting, but my guess is it's going to be callback-based.

**Comment by j-mueller** [comment]
@raduom wrote

> I was thinking of merging balance + fee management + signing + sending in one method

That won't work because for one the signing is done by a different process. So the txn would go to the wallet to be balanced & have fees added, then to the signing process, then back to the wallet which submits it to the blockchain. Sending the txn around is part of what the SCB does. 

For the mock wallet this means we need a separate HTTP endpoint for signing transactions, as well as one for submitting transactins to the chain.

@michaelpj  wrote

> I think it would be cleaner to have one id per hook, rather than one per "set of hooks".

I agree

> We should also check with the wallet folks what kind of interface they're expecting, but my guess is it's going to be callback-based.

I think it's going to be polling for the first iteration (this is what I told @raduom so that explains the proposed interface). I'll confirm in the call tomorrow.

**Comment by michaelpj** [comment]
> I think it's going to be polling for the first iteration

Ah okay, then ignore me. In that case, what about calling the function `poll`/`pollListeners`?

(Incidentally, I know we've been calling these things "hooks", but we might be better off with "listeners" or "handlers" which I think is more standard.)

**Comment by raduom** [comment]
Suggested API:

```
listen :: ListenerId
       -> Slot
       -- ^ The slot where we start listening for the Hook.
       -> Hook
       -- ^ The hooks are: watch address and wait for slot
       -> ()

poll :: ListenerId
     -> Slot
     -- ^ The slot where we start streaming events.
     -> Either Slot (Event, Slot)
     -- ^ There may be no matching event up to the current slot
     --   (left branch of the Either) or we have an event
     --   that needs processing, so we suspend streaming
     --   at (snd . fromRight)

balanceTx :: UnbalancedTx
          -> Error String Tx
          -- ^ We may be unable to balance the transaction.

sendTx :: Tx
       -> Error String ()
       -- ^ We may be unable to send this transaction.

currentSlot :: Slot
```

**Comment by raduom** [comment]
I've made a new iteration of the API.

* Renamed `ContractId` to `ListenerId`
* Renamed `runBlocks` to `poll` (given the first type is `ListenerId`, it reads close to Michael's suggestion of `pollListener` I think, and it is smaller. However, we can change the names however we see fit). 
* Added a `sendTx` (I was unaware that the signing process was separate).
* Note: This API should be able to support *only* the new contract API (this follows from conversations with Jann).

**Comment by michaelpj** [comment]
I think `listen` should be:
```
listen :: Slot
       -- ^ The slot where we start listening for the Hook.
       -> Hook
       -- ^ The hooks are: watch address and wait for slot
       -> m ListenerId`
```
i.e. the wallet backend should get to create ids.

We need a way to unregister listeners. We also possibly need a way to query them, or some notion of a session. Otherwise if the wallet crashes without unregistering its listeners, it may just re-register them when it comes back up, resulting in leakage.

That said, if we had a session, we would probably handle keeping it alive by sending keep-alive messages, and deregister its listeners when we hadn't heard from it in a while. We could just implement the same thing at the listener level, and say that if you haven't polled a listener for long enough it dies, but perhaps that's annoying.

As for `poll`, what about:
```
poll :: ListenerId
     -> SlotRange
     -- ^ The slot where we start streaming events.
     -> m [Event]
```
i.e. it asks for a range of slots (this can therefore include gradually polling segments of the past), and it gets back a list of events in order instead of just the first one. If the tail end of the slot range is later than the present, it could either give an error or just block until we get to that slot.

Or it could give you back some kind of fancy stream of events, and then asking for an unbounded slot range would be equivalent to the callback approach. Probably too complicated, but maybe worth thinking about.

IMO everything should just be in some monadic context that includes errors, so the signature of `sendTx` should just be:
```
sendTx :: Tx -> m ()
```

I still think `balanceTx` does not belong in this since this is the wallet *backend* API. The wallet backend doesn't know who the user is or what funds belong to them so it can't balance transactions.

**Comment by raduom** [comment]
Note: While I use Haskell syntax for functions, this is meant to be a REST endpoint API. By `m ()` you mean to consider the fact that there may be errors?

Also, returning a `[Event]` is not very useful as the contract may update the list on hooks on each Event, so what we listen for may change whenever an Event is received.

How about this:

```

**Comment by michaelpj** [comment]
> Note: While I use Haskell syntax for functions, this is meant to be a REST endpoint API. By m () you mean to consider the fact that there may be errors?

Yes, that's what I meant. I was confused because you had separate error handling for balance and submit, whereas I'd just shove that into the monad. That said, we obviously do want to distinguish between client recoverable errors like "transaction didn't validate" vs e.g. transport layer errors.

**Comment by j-mueller** [comment]
Ok so the call this morning with @rhyslbw and @mchakravarty clarified a few things:

1. We don't need to use polling for the blockchain events after all (my mistake, sorry for the confusion). We can do a subscription instead.
2. The signing of transactions is done by a separate process, as stated before. But the following points I had not considered until now. 
  (a) the signing process runs on the client, ie. in Javascript land - so we need to make sure that it comiles with GHCJS (the tricky part is the `cardano-crypto` package. For the mockchain we only need ED25519 for signing, and SHA2 / SHA3 256 for signatures. So the "mock signing process" could use an external JS implementation of those algorithms, via the JS FFI)
  (b) Coin selection probably needs to happen in the same place as the signing. It's true that in theory you only need to know the public keys to do coin selection, but in practice new private/public keys pairs are generated all the time and so it makes sense to do the coin selection in the process that knows all the current keys. (in the mockchain each wallet has exactly one private/public key pair which is not realistic)

That means the mock wallet (native Haskell .exe) is only responsible for

* Submitting signed and balanced transactions to the mockchain
* Registering event triggers and informing clients when mockchain events happen

The mock signing process (JS library produced by GHCJS) will be responsible for

* balancing (incl. coin selection) and signing transactions

I'll make a separate issue for the signing process.

**Comment by michaelpj** [comment]
> That means the mock wallet (native Haskell .exe) is only responsible for

Mock wallet *backend* right? Sorry to keep going on about this, but I at least keep getting those two muddled.

I'm a bit confused about what is in this "signing process". I thought it would essentially just be a "keyring", i.e. a hardened piece of code that has one job which is to look after the keys and not leak them, and occasionally sign things. But if it's doing coin selection then it *also* needs access to a whole bunch of the wallet's state too...

**Comment by michaelpj** [comment]
Also, the more I think about this the more I worry about what happens when either or both of the wallet and the wallet backend crash. There must be existing approaches to this kind of callback/subscription/listener problem that handle this, we should poke around.

**Comment by j-mueller** [comment]
> Mock wallet backend right? 

Yes that's right. I'll call it blockchain client from now on as suggested in slack

> I thought it would essentially just be a "keyring", i.e. a hardened piece of code that has one job which is to look after the keys and not leak them, and occasionally sign things.

Yeah that was my understanding as well but it's not the case

> There must be existing approaches to this kind of callback/subscription/listener problem that handle this, we should poke around.

Like a message bus? I always assumed that the SCB would take that role.

**Comment by michaelpj** [comment]
> Like a message bus? I always assumed that the SCB would take that role.

I honestly don't know. But here are some bad things that could happen:

- Wallet crashes, and either never comes up or comes back up and forgets about a listener, blockchain client keeps listeners indefinitely. Result: blockchain client keeps accumulating information for that listener forever.
- Blockchain client crashes and forgets about a listener. Wallet queries it later, gets error, no way (?) to get data that was missed while client was down/listener was forgotten.

Thinking about it, it would be really nice if we *could* have a stateless operation like Radu suggets:
```
poll :: SlotRange -> [Hook] -> m Event
```

If that works, then we can have a quasi-stateless solution:
- The wallet can tell the blockchain client about information it may want: this allows the blockchain client to fetch it incrementally and ahead of time.
- However, the wallet will always get the information it asks for, it might just be slower if it hasn't told the blockchain client to listen for it.

This way our state just affects how fast the blockchain client  can respond, so the worst case is performance degradation instead of incorrect behaviour.

**Comment by mchakravarty** [comment]
Signing can not be in the node client (which may run on a server distinct from the client device of the end user UI/app) for a variety of reasons:

* It may require a hardware dongle (which is with the user and only accessible on the client device).
* We want it air gapped from the node client, which will typically run on the same machine as the untrusted contract code.
* One node client may serve multiple end user devices.

All the issues related to dealing with crashes, restarting services etc is something the Application Platform needs to handle IMHO as it does orchestrate the system.

**Comment by raduom** [comment]
Suggested API:

```
data QueryResult = NoMatchFound
                   -- ^ No match was found when searching in SlotRange
                 | Synchronizing
                   -- ^ There is no data available, but the node client has
                   --   started to sync the requested data (try again later)
                 | FoundMatch [Event]


poll :: Maybe SlotRange
     -- ^ An optional slot range where we start polling events.
     -> Address
     -> m QueryResult

sendTx :: Tx
       -> m ()
       -- ^ We may be unable to send this transaction.

currentSlot :: Slot

```

Changes:

- removed signing and balancing (which I understand will be done by the signing process).
- removed the Hook and kept only watching addresses (the application platform can use currentSlot for polling the slot number)
- added the QueryResult data type to mirror my (our) understanding of how the wallet currently works (it can fetch historical data about addresses we care about, and, optionally, limit the data to a slot range)

**Comment by j-mueller** [comment]
`poll` looks OK to me.

I think we should be clearer about the `Event` type though. In the current implementation `Event = Ledger.Tx` - the whole transaction that touched an output at the address. But the contracts just use the transaction to construct the UTXO set at the address, discarding all the other information. So maybe it's enough to say `Event = [(Ledger.TxOut, Ledger.TxOutRef)]` (we need both the output and the outref to spend the input), and then `poll` would return the UTXO set at the end of the slot range (or at the current time if it's unbounded).

**Comment by michaelpj** [comment]
- I don't see a reason to have `NoMatchFound` as distinct from "here's a stream of events but it's empty".
- I don't like `Synchronizing`. What does a user actually do in response? Just wait for some unspecified period of time? At the very least, it should be a `Future` or some kind of waitable thing so it can get notified when it's done, but I think it would really be nice for it to just return a stream of events.
    - Waiting on the stream does the right thing if the wallet is still synchronizing.
    - We can easily start processing the first event when we get it.
    - We probably want some way to send a cancellation request if we timeout or we've got enough data.
    - We can build a streaming interface on top of a simpler one if that's what the real implementation wants, so I think we might as well use it for our mock version.

> I think we should be clearer about the Event type though. In the current implementation Event = Ledger.Tx - the whole transaction that touched an output at the address. But the contracts just use the transaction to construct the UTXO set at the address, discarding all the other information. 

I think it's fine for our low-level API to get the whole transaction. We might care about other things in the future, so I think this makes sense and we can build other stuff on top of it.

Consider e.g. a contract that waits for person X to pay to a script address: then it's not enough to see the UTXO set, you also want to see how it was made.

**Comment by raduom** [comment]
* `NoMatchFound`. I wanted to clarify this as an option. In hindsight, I agree we should remove it.
* There is something I do not understand here. The specification is for a REST endpoint, that will probably talk to a javascript client. Are you saying you would like a web socket here that maintains a connection and feeds the data to the client?
* I think that our low-level API should model the real application (or a future version thereof). If other people write contracts against the playground, they will (reasonably) expect them to run on the real chain as well. So we should check with the wallet if they can and will provide this data to us.

**Comment by michaelpj** [comment]
> The specification is for a REST endpoint, that will probably talk to a javascript client.

Hm, I was thinking this is the *functional* interface for the node client, but I'd forgotten that we were actually planning to write this to be used by the app platform until the node client catches up. Hmm.

In that case I guess we either implement the streaming ourselves on top of some kind of repeated querying interface, or we see if we can use some kind of streaming HTTP API. We still don't want to make it too hard for the app platform to do its job.

> I think that our low-level API should model the real application (or a future version thereof). If other people write contracts against the playground, they will (reasonably) expect them to run on the real chain as well. So we should check with the wallet team if they can and will provide this data to us.

My understanding from the call was that the wallet gets precisely the transactions that pay to that address. So passing that information on directly seems like the simplest thing, which is what I was suggesting.

**Comment by rhyslbw** [comment]
> That means the mock wallet (native Haskell .exe) is only responsible for
Submitting signed and balanced transactions to the mockchain
Registering event triggers and informing clients when mockchain events happen

I'd suggest not scoping this service to being a _mock_. The interface with the node will change as part of integration, but that's just standard progression of the software to support anything other than the mockchain. We will get a lot of value thinking about this service as a first-class _node client_

> Registering event triggers and informing clients when _**chain**_ events happen

This is the primary feature, and what I think the _node client_ should be named after. In my early models I had this service termed _**chain subscriber**_.

> Submitting signed and balanced transactions to the _**network**_

`cardano-wallet` is implementing this feature for Byron/Shelley now, but should be identical in the Goguen era codebase. The Haskell node design says that common functionality will be lifted to core libraries for clients to access, so best to align with @KtorZ and apply what's already in development. It makes sense for the new client to expose this API, but it's also conceivable in a scaled out system for the submission service to be separate since it's stateless (unlike the _chain subscriber_) and can be treated as ephemeral.

**Subscription design**
GraphQL subscriptions might be worth considering here to remove a lot of the implementation decisions, and result in conformance with an emerging common approach for push-based APIs

https://graphql.github.io/graphql-spec/June2018/#sec-Subscription

This approach puts the emphasis on defining the API, with transports then layered in for the particular use-case.

A good read to show how Apollo, a leading JS implementation has handled this:
https://github.com/apollographql/graphql-subscriptions
https://github.com/apollographql/graphql-subscriptions#pubsub-implementations
